market/deal_ingest_snap.go

```diff
@@ -309,16 +314,47 @@ func (p *PieceIngesterSnap) AllocatePieceToSector(ctx context.Context, maddr add
 		}
 
 		if len(candidates) == 0 {
-			return false, xerrors.Errorf("no suitable sectors found")
+			minEpoch := piece.DealSchedule.EndEpoch
+			maxEpoch := abi.ChainEpoch(maxExpiration)
+
+			minEpochDays := (minEpoch - head.Height()) / builtin.EpochsInDay
+			maxEpochDays := (maxEpoch - head.Height()) / builtin.EpochsInDay
+
+			return false, xerrors.Errorf("no suitable sectors found, minEpoch: %d, maxEpoch: %d, minExpirationDays: %d, maxExpirationDays: %d", minEpoch, maxEpoch, minEpochDays, maxEpochDays)
 		}
 
 		// todo - nice to have:
-		//  * double check the sector expiration
 		//  * check sector liveness
 		//  * check deadline mutable
 
 		candidate := candidates[0] // this one works best
 
+		si, err := p.api.StateSectorGetInfo(ctx, p.miner, abi.SectorNumber(candidate.Sector), types.EmptyTSK)
+		if err != nil {
+			return false, xerrors.Errorf("getting sector info: %w", err)
+		}
+
+		sectorLifeTime := si.Expiration - head.Height()
+		if sectorLifeTime < 0 {
+			return false, xerrors.Errorf("sector lifetime is negative!?")
+		}
+		if piece.DealSchedule.EndEpoch > si.Expiration {
+			return false, xerrors.Errorf("sector expiration is too soon: %d < %d", si.Expiration, piece.DealSchedule.EndEpoch)
+		}
+		if maxExpiration != 0 && si.Expiration > abi.ChainEpoch(maxExpiration) {
```

Comment:
This is too hard. We have 90 days grace period. We should not reject those sectors.

Reply:
See above. This is exactly as loose as it can be. maxExpiration is the highest allowable sector expiration based on termmax